engine/crates/composition/src/compose/object.rs

```diff
@@ -216,3 +216,28 @@ fn collect_overrides(fields: &[FieldWalker<'_>], ctx: &mut Context<'_>) -> Vec<f
 
     overrides
 }
+
+pub(crate) fn validate_shareable_object_fields_match(
+    definitions: &[DefinitionWalker<'_>],
+    ctx: &mut ComposeContext<'_>,
+) {
+    let all_fields: BTreeSet<StringId> = definitions
+        .iter()
+        .flat_map(|def| def.fields())
+        .filter(|field| !field.directives().inaccessible())
+        .map(|field| field.name().id)
+        .collect();
+
+    for field in all_fields {
+        for definition in definitions {
```

Comment:
I'd like to avoid doing (field.len() × definitions.len()) BTree lookups here, but there's no super easy fix at hand. Suggestions welcome.

Reply:
I think the best would be to do the outer loop by definition, have `all_fields` be a sorted array of fields, and iterate over the definitions fields. The definition fields would come in sorted order, so you can just move forward an index in the `all_fields` array until you find the matching field. It should also be more friendly with regards to caching.

Reply:
I'll try!